lib_test/bap_future/test_stream.ml

```diff
@@ -0,0 +1,565 @@
+  let f = Stream.last_before event ss last in
+  let max = List.length values in
+  fulfill_after_signals promise ~after:event_after ~max signal;
+  assert_equal ~ctxt expect (Future.peek_exn f)
+
+let before ctxt = 
+  let ss,signal, _ = Stream.of_list values in
+  let event, promise = Future.create () in
+  let f = Stream.before event ss in
+  Signal.repeat signal ~times:4 ();
+  Promise.fulfill promise ();
+  let expect = [0;1;2;3;] in
+  assert_equal ~ctxt expect (Future.peek_exn f)
+
+let suite =  
+  "Future_Stream" >:::
+  [
+    "send signal"       >:: send_signal;
+    "repeat signal: 0"  >:: repeat_signal ~times:0;
+    "repeat signal: 4"  >:: repeat_signal ~times:4;
+    "create"            >:: create;
+    "from"              >:: from;
+    "unfold"            >:: unfold;
+    "unfold_until"      >:: unfold_until;
+    "unfold'"           >:: unfold';
+    "repeat"            >:: repeat 42 ~times:3;
+    "repeat"            >:: repeat 42 ~times:0;
+    "of_list"           >:: of_list;
+    "of_array"          >:: of_array;
+    "of_sequence"       >:: of_sequence;
+    "watch"             >:: watch;
+    "observe"           >:: observe;
+    "subscribe"         >:: subscribe;
+    "unsubscribe"       >:: unsubscribe;
+    "on_wait"           >:: on_wait;
+    "on_subscribe"      >:: on_subscribe;
+    "on_unsubscribe"    >:: on_unsubscribe;
+    "has_subscribers"   >:: has_subscribers;
+    "diff subscribe id" >:: different_subscribe_id;
+    "diff watch id"     >:: different_watch_id;
+    "same watch id"     >:: same_watch_id;
+    "map'"              >:: map';
+    "map"               >:: map;
+    "apply"             >:: apply;
+    "split"             >:: split;
+    "unzip"             >:: unzip;
+    "once"              >:: once;
+    "parse"             >:: parse;
+    "before"            >:: before;
+    "hd"                >:: hd values;
+    "hd empty"          >:: hd [];
+    "tl"                >:: tl;  
+    "find existed"      >:: find values ~f:(fun x -> x = 3) ~expect:(Some 3);
+    "find not existed"  >:: find values ~f:(fun x -> x = 10) ~expect:None;
+    "take 3"            >:: take ~values ~expect:[0;1;2] ~how_match:3;
+    "take 0"            >:: take ~values ~expect:[] ~how_match:0;
+    "nth: n=3"          >:: nth values ~expect:(Some 3) ~n:3;
+    "nth: n=10"         >:: nth values ~expect:None ~n:10;
+    "upon: after 3"     >:: upon values ~expect:(Some 3) ~after:3;
+    "upon: after 10"    >:: upon values ~expect:None ~after:10;
+
+    "zip diff length"   >:: 
+    zip ~values:[0;1;2;3;] ~values':[0; 1; 2; 3; 4; 5;] 
+      ~expect:[0,0; 1,1; 2,2; 3,3;];
+
+    "filter_map"        >:: 
+    filter_map ~values:[0;1;2;3;4;5;6;7;8] ~expect:[0;2;4;6;8] ~f:even_opt;
+
+    "merge diff length" >:: 
+    merge ~values:[0;1;2;3;] ~values':[0.; 1.; 2.; 3.; 4.; 5.; ]
+      ~expect:[0.5; 2.5; 4.5; 6.5] ~f:(fun x y -> float x +. y +. 0.5);
+
+    "foldw stride=1 width=3" >::
```

Comment:
add tests for `foldw` with function `fun xs x -> x :: xs` with widths ranging from 0 to 10 and strides ranging from 0 to 10.